lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -74,69 +118,9 @@ sub ajax {
-                my $hash     = {
-                    id        => $n->id,
-                    directory => $n->directory->name,
-                    filename  => $filename,
-                };
-                my $needle_dir_path = $n->directory->path;
-                my $real_needle_dir_id;
-
-                if ($needle_dir_real_paths{$needle_dir_path}) {
-                    my $real_path = $needle_dir_real_paths{$needle_dir_path}->{real_path};
-                    $real_needle_dir_id = $needle_dir_real_paths{$needle_dir_path}->{real_path_id}
-                      if $real_path && $needle_dir_path ne $real_path;
-                }
-                else {
-                    my $real_path_id         = $n->directory->id;
-                    my $needle_dir_real_path = realpath($needle_dir_path);
-                    if ($needle_dir_real_path && $needle_dir_real_path ne $needle_dir_path) {
-                        my $real_needle_dir
-                          = $self->schema->resultset('NeedleDirs')->find({path => $needle_dir_real_path});
-                        if ($real_needle_dir) {
-                            $real_needle_dir_id = $real_needle_dir->id;
-                            $real_path_id       = $real_needle_dir_id;
-                        }
-                    }
-                    $needle_dir_real_paths{$needle_dir_path} = {
-                        real_path    => $needle_dir_real_path,
-                        real_path_id => $real_path_id,
-                    };
-                }
-                my $real_needle_info;
-                $real_needle_info
-                  = $schema->resultset('Needles')->find({dir_id => $real_needle_dir_id, filename => $filename})
-                  if $real_needle_dir_id;
-                $n = $real_needle_info if $real_needle_info;
-
-                $hash->{last_seen}  = $n->last_seen_time    || 'never';
-                $hash->{last_match} = $n->last_matched_time || 'never';
-
-                if (my $last_seen_module_id = $n->last_seen_module_id) {
-                    $hash->{last_seen_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_seen_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                if (my $last_matched_module_id = $n->last_matched_module_id) {
-                    $hash->{last_match_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_matched_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                push(@data, $hash);
-            }
-            return \@data;
+            my $needle_dirs = $self->schema->resultset('NeedleDirs');
+            my $needles     = $self->schema->resultset('Needles');
+            [map { $self->_prepare_data_table($_, {}, $needle_dirs, $needles) } shift->all];
```

Comment:
As @Martchus said,  the `{}` should be outside the loop.  It is used to store all needle dirs relative real path and id. so the value should be added when query a new needle dir which is not in this hash.

Reply:
I understand that. I thought I crosschecked this with tests but I tested again, e.g. by moving the hash *within* `_prepare_data_table` expecting that tests would fail but unfortunately the would not. So I understand that this change does not survive a simple mutation test but would probably only surface in a performance test. Also when checking with NYTProf regardless of what I do, calls to "realpath" are always 14 times in t/ui/21-admin-needles.t and take accumulated 279µs 
. Anyway I changed this to be more explicit.